File: lib/vying/parts/board/plugins/custodial_capture.rb
# frozen_string_literal: true

# Copyright 2007, Eric Idema except where otherwise noted.
# You may redistribute / modify this file under the same terms as Ruby.

require 'vying'

# Adds the #will_flip? and #flip methods to Board.  These methods can be used
# for implementing Othello and like games.

module Board::Plugins::CustodialCapture

  # Place the given piece p at Coord c and make any valid custodial captures
  # as a side effect.  To limit the captures to a certain number of pieces,
  # provide a range.  To make unlimited captures the range can be omitted.
  #
  # For example:
  #
  #   b.custodial_capture( Coord[:a1], :x, 3..4 )
  #
  # With the given lines (where :a1 is represented with a '.'):
  #
  #     .oox    => no capture
  #     .ooox   => capture
  #     .oooox  => capture
  #     .ooooox => no capture
  #
  # This method returns an array containing the coords of pieces that were
  # successfully captured.  It is possible that no pieces will be capture,
  # in which case p is still placed at coord c.

  def custodial_capture(c, p, range=nil)
    custodial(c, p, nil, range)
  end

  # Will playing piece p at coord c result in a custodial capture?  The
  # optional range can be used to restrict the capture test.

  def custodial_capture?(c, p, range=nil)
    a = directions.zip(coords.neighbors_nil(c))
    a.each do |d, nc|
      next if self[nc].nil? || self[nc] == p

      bt = [nc]
      while bt << coords.next(bt.last, d)
        break if self[bt.last].nil?
        break if range && bt.length - 1 < range.first && self[bt.last] == p
        next  if range && bt.length - 1 < range.first
        break if range && bt.length - 1 > range.last

        return true if self[bt.last] == p
      end
    end

    false
  end

  # Does the heavy lifting for custodial actions.  Takes the coord to play,
  # the piece to place, what to replace captured pieces with, and the range
  # to effect.

  def custodial(c, p, replacement=nil, range=nil)
    self[c] = p

    cap = []
    a = directions.zip(coords.neighbors_nil(c))
    a.each do |d, nc|
      next if self[nc].nil? || self[nc] == p

      bt = [nc]
      while bt << coords.next(bt.last, d)
        break if self[bt.last].nil?
        break if range && bt.length - 1 < range.first && self[bt.last] == p
        next  if range && bt.length - 1 < range.first
        break if range && bt.length - 1 > range.last

        next unless self[bt.last] == p
        bt.pop
        cap += bt
        break
      end
    end

    cap.each do |cc|
      self[cc] = replacement
    end

    cap
  end

  private :custodial

end

File: ext/vying/c/parts/board/custodial_capture.c
/* Copyright 2007, Eric Idema except where otherwise noted.
 * You may redistribute / modify this file under the same terms as Ruby.
 */

#include "ruby.h"
#include "board.h"

/*
 *  Tests whether placing the given piece at the given coord would result
 *  in pieces being flipped any of the Board#directions.
 *
 *  call-seq:
 *    custodial_capture?( coord, piece, range=nil ) -> boolean
 *
 */

VALUE custodial_capture_valid( int argc, VALUE *argv, VALUE self ) {
  int x = NUM2INT(rb_funcall( argv[0], id_x, 0 ));
  int y = NUM2INT(rb_funcall( argv[0], id_y, 0 ));

  VALUE p = argv[1];

  VALUE range = Qnil;
  int first = 0;
  int last = 0;

  VALUE dir = rb_iv_get( self, "@directions" );

  int i, blen;

  if( argc > 2 && RTEST(argv[2]) ) {
    range = argv[2];
    first = NUM2INT(rb_funcall( range, id_first, 0 ));
    last  = NUM2INT(rb_funcall( range, id_last,  0 ));
  }

  for( i = 0; i < RARRAY_LEN(dir); i++ ) {
    VALUE d = rb_ary_entry( dir, i );
    int dx = 0, dy = 0, nx, ny;
    VALUE np;

    blen = 0;

    if( d == sym_n ) {
      dx = 0;
      dy = -1;
    }
    else if( d == sym_s ) {
      dx = 0;
      dy = 1;
    }
    else if( d == sym_w ) {
      dx = -1;
      dy = 0;
    }
    else if( d == sym_e ) {
      dx = 1;
      dy = 0;
    }
    else if( d == sym_ne ) {
      dx = 1;
      dy = -1;
    }
    else if( d == sym_nw ) {
      dx = -1;
      dy = -1;
    }
    else if( d == sym_se ) {
      dx = 1;
      dy = 1;
    }
    else if( d == sym_sw ) {
      dx = -1;
      dy = 1;
    }

    nx = x+dx;
    ny = y+dy;

    blen++;

    np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    if( NIL_P(np) || np == p ) {
      continue;
    }

    nx += dx;
    ny += dy;
    np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    while( np != Qnil ) {
      if( range != Qnil && blen > last ) {
        break;
      }
      if( range != Qnil && blen < first && np == p ) {
        break;
      }
      if( np == p && ! (range != Qnil && blen < first) ) {
        return Qtrue;
      }

      nx += dx;
      ny += dy;

      blen++;

      np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    }
  }

  return Qfalse;
}

/*
 *  Place the given piece at the given coord.  Replace any pieces that would
 *  be victims of custodial capture in one of Board#directions.
 *
 *  call-seq:
 *    custodial( coord, piece, replacement=nil, range=nil ) -> [captured coords]
 *
 */

VALUE custodial( int argc, VALUE *argv, VALUE self ) {
  int x = NUM2INT(rb_funcall( argv[0], id_x, 0 ));
  int y = NUM2INT(rb_funcall( argv[0], id_y, 0 ));

  VALUE p = argv[1];

  VALUE replacement = Qnil;

  VALUE range = Qnil;
  int first = 0;
  int last = 0;

  VALUE dir = rb_iv_get( self, "@directions" );

  VALUE cap = rb_ary_new();

  int w = NUM2INT(rb_iv_get( self, "@width" ));
  int h = NUM2INT(rb_iv_get( self, "@height" ));

  int bt[(w > h) ? w : h][2];
  int blen = 0;

  int i;

  if( argc > 2 ) {
    replacement = argv[2];
  }

  if( argc > 3 && RTEST(argv[3]) ) {
    range = argv[3];
    first = NUM2INT(rb_funcall( range, id_first, 0 ));
    last  = NUM2INT(rb_funcall( range, id_last,  0 ));
  }

  for( i = 0; i < RARRAY_LEN(dir); i++ ) {
    VALUE d = rb_ary_entry( dir, i );
    int dx = 0, dy = 0, nx, ny;
    VALUE np;

    blen = 0;

    if( d == sym_n ) {
      dx = 0;
      dy = -1;
    }
    else if( d == sym_s ) {
      dx = 0;
      dy = 1;
    }
    else if( d == sym_w ) {
      dx = -1;
      dy = 0;
    }
    else if( d == sym_e ) {
      dx = 1;
      dy = 0;
    }
    else if( d == sym_ne ) {
      dx = 1;
      dy = -1;
    }
    else if( d == sym_nw ) {
      dx = -1;
      dy = -1;
    }
    else if( d == sym_se ) {
      dx = 1;
      dy = 1;
    }
    else if( d == sym_sw ) {
      dx = -1;
      dy = 1;
    }

    nx = x+dx;
    ny = y+dy;

    np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    if( NIL_P(np) || np == p ) {
      continue;
    }

    bt[blen][0] = nx;
    bt[blen][1] = ny;
    blen++;    

    nx += dx;
    ny += dy;
    np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    while( np != Qnil ) {
      if( range != Qnil && blen > last ) {
        break;
      }
      if( range != Qnil && blen < first && np == p ) {
        break;
      }
      if( np == p && ! (range != Qnil && blen < first ) ) {
        int j;
        for( j = 0; j < blen; j++ ) {
          VALUE cx = INT2NUM(bt[j][0]);
          VALUE cy = INT2NUM(bt[j][1]);

          rb_funcall( self, id_set, 3, cx, cy, replacement );
          rb_ary_push( cap, rb_funcall( Coord, id_new, 2, cx, cy ) );
        }

        break;
      }

      bt[blen][0] = nx;
      bt[blen][1] = ny;
      blen++;    

      nx += dx;
      ny += dy;
      np = board_get( self, INT2NUM(nx), INT2NUM(ny) );
    }
  }

  board_set( self, INT2NUM(x), INT2NUM(y), p );

  return cap;
}

